frontend/src/lib/ReportNode.ts

```diff
@@ -281,10 +282,17 @@ export class ElementNode implements ReportNode {
     element: Quiver,
     namedDataSet: ArrowNamedDataSet
   ): Quiver {
+    if (namedDataSet.hasName) {
+      throw new Error(
+        "Add rows cannot be used with a named dataset for this element."
+      )
+    }
+
     const newQuiver = new Quiver(namedDataSet.data as IArrow)
     element.addRows(newQuiver)
 
-    return element
+    // Clone deep is needed to update the React component.
+    return cloneDeep(element)
```

Comment:
Nit: can you make "clone" a public function on `Quiver`, and move the `cloneDeep` logic into that function and out of `ReportNode`, so that Quiver is entirely in charge of cloning? We may want to change the implementation down the road.

In particular, I could see us having a `Quiver.cloneShallow` function that would re-use the internal parsed Arrow data, which would be quite a bit faster than a deep clone:

```typescript
  public cloneShallow = (): Quiver => {
    const theClone = new Quiver()
    theClone._index = this._index
    theClone._columns = this._columns
    theClone._data = this._data
    theClone._types = this._types
    theClone._styler = this._styler
    return theClone
  }
```

^Actually, is this what we should be doing for `add_rows` already? If we don't mind mutating the source element, a shallow clone should get us what we need for React updating.

Reply:
Moved to https://github.com/streamlit/streamlit-issues/issues/162